cmd/utils.go

```diff
@@ -95,9 +96,92 @@ func splitStr(path, sep string, n int) []string {
 	return splits
 }
 
+// Parse scheme format, supported formats are
+// - "http"
+// - "http+s3v2"
+// - "http+s3v4"
+func parseScheme(scheme string) (string, string, *probe.Error) {
+	tokens := strings.Split(scheme, "+")
+	if len(tokens) > 2 || len(tokens) < 1 {
+		return "", "", errInvalidArgument().Trace(scheme)
+	}
+
+	httpScheme := tokens[0]
+	if httpScheme != "http" && httpScheme != "https" {
+		return "", "", errInvalidArgument().Trace(scheme, httpScheme)
+	}
+
+	signatureType := "s3v4"
+	if len(tokens) == 2 {
+		signatureType = tokens[1]
+	}
+	if !isValidAPI(signatureType) {
+		return "", "", errInvalidArgument().Trace(signatureType, scheme)
+	}
+
+	return tokens[0], signatureType, nil
+}
+
+func buildS3ConfigFromEnv(envConfig, urlStr string) (*Config, *probe.Error) {
+	_, p := url2Alias(urlStr)
+
+	// We have a valid alias in env. We populate the credentials from env.
+	s3Config := new(Config)
+
+	u, e := url.Parse(envConfig)
+	if e != nil {
+		return nil, probe.NewError(e).Trace(envConfig)
+	}
+	var (
+		accessKey    = u.User.Username()
+		secretKey, _ = u.User.Password()
+	)
+	if !isValidAccessKey(accessKey) || !isValidSecretKey(secretKey) {
+		return nil, probe.NewError(fmt.Errorf("Access/Secret keys %s "+
+			"are found in your environment but not suitable for use. "+
+			"Falling back to the standard config.", envConfig))
+	}
+
+	scheme, signatureType, err := parseScheme(u.Scheme)
+	if err != nil {
+		return nil, err.Trace(u.Scheme, envConfig)
+	}
+
+	hostURL, e := url.Parse(scheme + "://" + u.Host)
+	if e != nil {
+		return nil, probe.NewError(e).Trace(envConfig)
+	}
+
+	if !isValidHostURL(hostURL.String()) {
+		return nil, errInvalidURL(hostURL.String())
+	}
+
+	s3Config.HostURL = urlJoinPath(hostURL.String(), p)
+	s3Config.Signature = signatureType
+	s3Config.AccessKey = accessKey
+	s3Config.SecretKey = secretKey
+	s3Config.AppName = "mc"
+	s3Config.AppVersion = Version
+	s3Config.AppComments = []string{os.Args[0], runtime.GOOS, runtime.GOARCH}
+	s3Config.Debug = globalDebug
+	s3Config.Insecure = globalInsecure
+
+	return s3Config, nil
+}
+
+func getAlias(aliasURL string) string {
+	alias, _ := url2Alias(aliasURL)
+	return alias
+}
+
 // buildS3Config fetches config related to the specified alias
 // to create a new config structure
 func buildS3Config(alias, urlStr string) (*Config, *probe.Error) {
+	// Check if we can get the values from ENV, if not use from `config.json`.
+	if envConfig, ok := os.LookupEnv("MC_TARGET_" + getAlias(urlStr)); ok {
```

Comment:
I am not sure, isn't it easier to use 'alias' variable instead of 'getAlias(urlStr)' ?

Reply:
alias variable is empty @vadmeste if alias not present. So there is a catch 22 situation :-)